utils/math_test.go

```diff
@@ -0,0 +1,213 @@
+package utils_test
+
+import (
+	"fmt"
+	"math"
+	"testing"
+
+	sdkmath "cosmossdk.io/math"
+
+	"github.com/stretchr/testify/require"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/provlabs/vault/utils"
+)
+
+func TestCalculateSharesFromAssets(t *testing.T) {
+	denom := "vaultshare"
+
+	tests := []struct {
+		name        string
+		assets      sdkmath.Int
+		totalAssets sdkmath.Int
+		totalShares sdkmath.Int
+		expected    sdk.Coin
+		expectErr   bool
+	}{
+		{
+			name:        "first deposit (1:1 mapping)",
+			assets:      sdkmath.NewInt(100),
+			totalAssets: sdkmath.NewInt(0),
+			totalShares: sdkmath.NewInt(0),
+			expected:    sdk.NewCoin(denom, sdkmath.NewInt(100)),
+			expectErr:   false,
+		},
+		{
+			name:        "proportional minting",
+			assets:      sdkmath.NewInt(50),
+			totalAssets: sdkmath.NewInt(100),
+			totalShares: sdkmath.NewInt(200),
+			expected:    sdk.NewCoin(denom, sdkmath.NewInt(100)),
+			expectErr:   false,
+		},
+		{
+			name:        "rounding down",
+			assets:      sdkmath.NewInt(1),
+			totalAssets: sdkmath.NewInt(3),
+			totalShares: sdkmath.NewInt(10),
+			expected:    sdk.NewCoin(denom, sdkmath.NewInt(3)),
+			expectErr:   false,
+		},
+		{
+			name:        "zero asset input",
+			assets:      sdkmath.NewInt(0),
+			totalAssets: sdkmath.NewInt(100),
+			totalShares: sdkmath.NewInt(1000),
+			expected:    sdk.NewCoin(denom, sdkmath.NewInt(0)),
+			expectErr:   false,
+		},
+		{
+			name:        "zero totalAssets and totalShares, zero input",
+			assets:      sdkmath.NewInt(0),
+			totalAssets: sdkmath.NewInt(0),
+			totalShares: sdkmath.NewInt(0),
+			expected:    sdk.NewCoin(denom, sdkmath.NewInt(0)),
+			expectErr:   false,
+		},
+		{
+			name:        "extremely large values (1:1)",
+			assets:      sdkmath.NewIntFromUint64(math.MaxUint64),
+			totalAssets: sdkmath.NewIntFromUint64(math.MaxUint64),
+			totalShares: sdkmath.NewIntFromUint64(math.MaxUint64),
+			expected:    sdk.NewCoin(denom, sdkmath.NewIntFromUint64(math.MaxUint64)),
+			expectErr:   false,
+		},
+		{
+			name:        "very small asset compared to large totalAssets",
+			assets:      sdkmath.NewInt(1),
+			totalAssets: sdkmath.NewInt(1_000_000_000),
+			totalShares: sdkmath.NewInt(1_000_000_000),
+			expected:    sdk.NewCoin(denom, sdkmath.NewInt(1)), // (1 * 1e9 / 1e9)
+			expectErr:   false,
+		},
+		{
+			name:        "assets > totalAssets",
+			assets:      sdkmath.NewInt(2000),
+			totalAssets: sdkmath.NewInt(1000),
+			totalShares: sdkmath.NewInt(1000),
+			expected:    sdk.NewCoin(denom, sdkmath.NewInt(2000)), // (2000 * 1000 / 1000)
+			expectErr:   false,
+		},
+		{
+			name:        "assets < totalAssets",
+			assets:      sdkmath.NewInt(100),
+			totalAssets: sdkmath.NewInt(1000),
+			totalShares: sdkmath.NewInt(1000),
+			expected:    sdk.NewCoin(denom, sdkmath.NewInt(100)), // (100 * 1000 / 1000)
+			expectErr:   false,
+		},
```

Comment:
we should add a few more test cases for invalid or negative numbers and their expected error conditions to fully cover the range of input values our method can accept even if they are not valid.